[trino-ignite] Ensure primary_key columns exist while creating table

sahoss opened the pull request:
## Description

Adds a check to validate that the column names defined in primary_key table property exists as column names in the table definition.

Fixes #16215

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

sahoss commented on plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteClient.java:
```diff
@@ -59,11 +67,36 @@
             new BaseJdbcConfig(),
-            session -> { throw new UnsupportedOperationException(); },
+            new TestRetryingConnectionFactory.MockConnectorFactory(
+                    TestRetryingConnectionFactory.MockConnectorFactory.Action.RETURN),
```
need this change as there is 1 call to openConnection(session) which was failing without this mock.

sahoss commented:
@chenjian2664  @ebyhr please take a look :)

chenjian2664 commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -391,13 +392,21 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+        for (String primaryKey : primaryKeys) {
+            if (!columnNames.contains(primaryKey)) {
+                String message = format("%s specified as PRIMARY KEY is not present in columns", primaryKey);
+                throw new TrinoException(COLUMN_NOT_FOUND, message);
```
`INVALID_TABLE_PROPERTY` error code would be better

chenjian2664 commented on plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteClient.java:
```diff
@@ -59,11 +67,36 @@
             RemoteQueryModifier.NONE);
 
+    @Test
+    public void testBeginCreateTableWithNonExistingPrimaryKey()
```
Could you add test into `TestIgniteConnectorTest` instead of here.

chenjian2664 commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -391,13 +392,21 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         List<String> primaryKeys = IgniteTableProperties.getPrimaryKey(tableProperties);
         checkArgument(primaryKeys.size() < columns.size(), "Ignite table must have at least one non PRIMARY KEY column.");
+
+        for (String primaryKey : primaryKeys) {
```
It would be better to move the logic into the loop `for (Map.Entry<String, Object> propertyEntry : tableProperties.entrySet()) {`, we might support more properties latter.

sahoss commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -391,26 +392,36 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
-            columnDefinitions.add(quoted(IGNITE_DUMMY_ID) + " VARCHAR NOT NULL");
-            primaryKeys = ImmutableList.of(IGNITE_DUMMY_ID);
-        }
-        columnDefinitions.add("PRIMARY KEY (" + join(", ", primaryKeys.stream().map(this::quoted).collect(joining(", "))) + ")");
 
         for (Map.Entry<String, Object> propertyEntry : tableProperties.entrySet()) {
```
have changed the relative order between the `if (primaryKeys.isEmpty())` block and `for (Map.Entry<String, Object> propertyEntry : tableProperties.entrySet()) {` 
the latter only performs validation checks and this reorder doesnt affect correctness..

reordering is needed because without it, the `if (primaryKeys.isEmpty())` adds a dummy id column and will require special case while checking for primary key present in columnsName list.

ebyhr commented on plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:
```diff
@@ -147,6 +147,13 @@ public void testCreateTableWithCommaPropertyColumn()
+    {
+        assertQueryFails("CREATE TABLE test_create_table_non_existing_primary_key (a bigint, b double) with (primary_key = ARRAY['c'])",
+                "Column 'c' specified in property 'primary_key' doesn't exist in table");
+    }
```
Let's add below assertions to this test method:
* Specify existing column with uppercase in primary_key. e.g. A
* Specify `dummy_id` column explicitly in primary_key

sahoss replied:
> @sahoss Please feel free to ping me after addressing comments.

@ebyhr done.
deleted the unused branch in https://github.com/trinodb/trino/pull/16266/commits/328e5a650d773478fb17b99f608dd243804ae374

ebyhr replied:
Reminder.

sahoss replied:
@ebyhr this is done. please take a look to see if this was how you meant it?

```
assertQueryFails("CREATE TABLE test_create_table_non_existing_primary_key (a bigint, b bigint, c bigint) with (primary_key = ARRAY['A', 'b', 'dummy_id'])",
                "Column 'dummy_id' specified in property 'primary_key' doesn't exist in table")
```

ebyhr replied:
Please don't mix up the assertions.

sahoss replied:
@ebyhr for `Specify existing column with uppercase in primary_key. e.g. A`
i.e. 
```
CREATE TABLE test_create_table_non_existing_primary_key (a bigint, b bigint, c bigint) with (primary_key = ARRAY['A'])
```
is the expected behavior to fail the request? I see that we are lowercasing the column names(in the table definition), do we also lowercase the attributes in primary key property or do we fail the request?

sahoss replied:
@ebyhr 
`Specify dummy_id column explicitly in primary_key` 
is this a special case of sorts or does the current test case exercise this scenario? ptal at the previous question as well.

ebyhr replied:
I expect both scenarios fail.

sahoss replied:
@ebyhr please take a look. have added 2 separate failure assertions.

ebyhr replied:
`dummy_id` is a special naming in the connector. I would recommend splitting the 1st `assertQueryFails` into two assertions.

sahoss replied:
> `dummy_id` is a special naming in the connector. I would recommend splitting the 1st `assertQueryFails` into two assertions.

@ebyhr done. please take a look.

ebyhr replied:
Could you update like below? 
```java
    @Test
    public void testCreateTableWithNonExistingPrimaryKey()
    {
        String tableName = "test_invalid_primary_key" + randomNameSuffix();
        assertQueryFails("CREATE TABLE " + tableName + "(a bigint) WITH (primary_key = ARRAY['not_existing_column'])",
                "Column 'not_existing_column' specified in property 'primary_key' doesn't exist in table");

        assertQueryFails("CREATE TABLE " + tableName + "(a bigint) WITH (primary_key = ARRAY['dummy_id'])",
                "Column 'dummy_id' specified in property 'primary_key' doesn't exist in table");

        assertQueryFails("CREATE TABLE " + tableName + "(a bigint) WITH (primary_key = ARRAY['A'])",
                "Column 'A' specified in property 'primary_key' doesn't exist in table");
    }
```

ebyhr commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -391,26 +392,36 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+            if (PRIMARY_KEY_PROPERTY.equalsIgnoreCase(propertyKey)) {
+                for (String primaryKey : primaryKeys) {
+                    if (!columnNames.contains(primaryKey)) {
+                        String message = format("Column '%s' specified in property '%s' doesn't exist in table", primaryKey, PRIMARY_KEY_PROPERTY);
```
nit: Inline the variable.

ebyhr commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -391,26 +392,36 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
 
         for (Map.Entry<String, Object> propertyEntry : tableProperties.entrySet()) {
             String propertyKey = propertyEntry.getKey();
             if (!PRIMARY_KEY_PROPERTY.equalsIgnoreCase(propertyKey)) {
                 throw new UnsupportedOperationException("Not support table property " + propertyKey);
             }
```
This branch seems reachable because it's ensured in engine. You can delete in a separated commit.

ebyhr commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -373,7 +374,7 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
             columnTypes.add(columnMetadata.getType());
         }
-        String sql = buildCreateSql(schemaTableName, columns.build(), tableMetadata.getProperties());
+        String sql = buildCreateSql(schemaTableName, columns.build(), columnNames.build(), tableMetadata.getProperties());
```
The both `columnNames` and `primaryKeys` are built twice. Could you change the arguments of `buildCreateSql` to `SchemaTableName schemaTableName, List<String> columns, List<String> primaryKeys`?

sahoss replied:
@ebyhr 
Did you mean like `private String buildCreateSql(SchemaTableName schemaTableName, List<String> columns, List<String> primaryKeys)`? 

if we do this, its a bit hard to check for existence of primaryKey in columns as columns contains columnName+type (eg: `name varchar(25)`).
we will need to pass either table properties or columnNames into buildCreateSql.

sahoss replied:
>The both columnNames and primaryKeys are built twic

Yes. we can avoid this step by calling build after 
```
for (ColumnMetadata columnMetadata : tableMetadata.getColumns()) {
            columns.add(getColumnDefinitionSql(session, columnMetadata, columnMetadata.getName()));
            columnNames.add(columnMetadata.getName());
            columnTypes.add(columnMetadata.getType());
        }
```

ebyhr replied:
You can check the column existence inside of `beginCreateTable()` instead.

sahoss replied:
done.

ebyhr reviewed (commented):
Thanks for your contribution. Please squash commits into one and follow this guideline for commit messages. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

By the way, have you already submit CLA?

sahoss commented:
> Thanks for your contribution. Please squash commits into one and follow this guideline for commit messages. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages
> 
:)  

Will do this
> By the way, have you already submit CLA?

Yes. i have submit it via email, but not heard back on the email.

sahoss commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -366,13 +368,18 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         for (ColumnMetadata columnMetadata : tableMetadata.getColumns()) {
             columns.add(getColumnDefinitionSql(session, columnMetadata, columnMetadata.getName()));
-            columnNames.add(columnMetadata.getName());
+            columnNamesBuilder.add(columnMetadata.getName());
```
please let me know, for consistency, if i can make columnames, columnTypes into builder as well and build it into a string variable.

sahoss commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -366,13 +368,18 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         int expectedSize = tableMetadata.getColumns().size();
         ImmutableList.Builder<String> columns = ImmutableList.builderWithExpectedSize(expectedSize);
-        ImmutableList.Builder<String> columnNames = ImmutableList.builderWithExpectedSize(expectedSize);
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builderWithExpectedSize(expectedSize);
```
please let me know, for consistency, if i can make columnames, columnTypes into builder as well and build it into a string variable.

ebyhr commented:
@sahoss Please feel free to ping me after addressing comments.

sahoss commented:
> @sahoss Please feel free to ping me after addressing comments.

@ebyhr Done. please take a look. 
https://github.com/trinodb/trino/pull/16266/commits/328e5a650d773478fb17b99f608dd243804ae374 deletes the unused branch.

https://github.com/trinodb/trino/pull/16266#discussion_r1117984428 and https://github.com/trinodb/trino/pull/16266#discussion_r1117982606 is addressed as well.

ebyhr commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -366,13 +367,18 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+        List<String> columnNames = columnNamesBuilder.build();
+
+        checkPrimaryKeyExistsForBeginCreateTable(columnNames, IgniteTableProperties.getPrimaryKey(tableMetadata.getProperties()));
+
```
You can check like this:
```java
        List<String> columnNames = columnNamesBuilder.build();
        List<String> primaryKeys = IgniteTableProperties.getPrimaryKey(tableMetadata.getProperties());
        for (String primaryKey : primaryKeys) {
            if (!columnNames.contains(primaryKey.toLowerCase(ENGLISH))) {
                throw new TrinoException(INVALID_TABLE_PROPERTY, format("Column '%s' specified in property '%s' doesn't exist in table", primaryKey, PRIMARY_KEY_PROPERTY));
            }
        }

        String sql = buildCreateSql(schemaTableName, columns.build(), primaryKeys);
...
                    primaryKeys.isEmpty() ? Optional.of(IGNITE_DUMMY_ID) : Optional.empty());
...
    private String buildCreateSql(SchemaTableName schemaTableName, List<String> columns, List<String> primaryKeys)
```

sahoss replied:
done

sahoss replied:
@ebyhr Have added the check in the above way.

Where do we want the ternary expression check?
```
 primaryKeys.isEmpty() ? Optional.of(IGNITE_DUMMY_ID) : Optional.empty());
```

ebyhr replied:
You can fix L399.

sahoss replied:
done

sahoss commented:
@ebyhr wondering if you could take a look at this pr (https://github.com/trinodb/trino/pull/16266#discussion_r1121263446 has been addressed)

also, `verification/cla-signed` bot is failing, but i have mailed my CLA. anything/anyone specifically i should do or reachout to? thanks.

ebyhr commented:
@cla-bot check

ebyhr reviewed (approved):
> Remove unreachable branch and for loop

I would change to "Remove unreachable code from IgniteClient"

sahoss commented:
@ebyhr updated the commit message and updated the test_name variable. ptal!

ebyhr commented on plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:
```diff
@@ -415,13 +415,6 @@ private String buildCreateSql(SchemaTableName schemaTableName, List<String> colu
         }
         columnDefinitions.add("PRIMARY KEY (" + join(", ", primaryKeys.stream().map(this::quoted).collect(joining(", "))) + ")");
 
-        for (Map.Entry<String, Object> propertyEntry : tableProperties.entrySet()) {
```
The commit order is wrong. `tableProperties` is already removed in the 1st commit.

sahoss replied:
@ebyhr thanks. i have fixed it. i verified by running tests on both the commits.
